File: Sliding_window/Java/Minimum Size Subarray Sum.java
// Time: O(n)
// Space: O(1)
// Problem link: https://leetcode.com/problems/minimum-size-subarray-sum/description/

class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int min = nums.length + 1;
        int sum = 0;
        int j = 0;
        for(int i = 0; i < nums.length; i++){
            sum += nums[i];
            while(sum >= target){
                min = Math.min(min,i - j + 1);
                sum -= nums[j++];
            }
        }
        return min == nums.length + 1 ? 0 : min;
    }
}

File: Sliding_window/Python/Minimum Size Subarray Sum.py
# Time: O(n)
# Space: O(1)
# Problem link: https://leetcode.com/problems/minimum-size-subarray-sum/description/
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        total = 0
        result = len(nums) + 1
        i = j = 0

        while j < len(nums):
            total += nums[j]
            while total >= target:
                result = min(result, j - i + 1)
                total -= nums[i]
                i += 1
            j += 1

        return result if result != len(nums) + 1 else 0
